### PR TITLE
[NOID] Fixes #2283: Utilize KernelTransaction.setStatusDetails() in 4.4

### LINKED ISSUE #2283: Utilize KernelTransaction.setStatusDetails() in 4.4
For procedures that have long running operations, like periodic iterate or import procedures, use the new functionality to report progress

* periodic.iterate - how many batches success / how many errors / how many 
* load csv/xml/json
* from document/import-json/import-csv  -> how much data processed (rows, elements) / how much data created in the graph

### PR DESCRIPTION
**Pull request by vga91**
Fixes #2283

I don't added `@Context public KernelTransaction kernelTx` due to https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/2265#discussion_r738180605.

I used `final KernelTransaction ktx = ((InternalTransaction) tx).kernelTransaction();` instead

### CODE REVIEW

**Comment by ncordon** [comment]
Can you replay the CI here, please?

File: full/src/main/java/apoc/load/LoadHtml.java [hunk 1]
```diff
@@ -114,11 +121,12 @@ public static List<Map<String, Object>> getElements(Elements elements, LoadHtmlC
                 if (conf.isChildren()) {
                     if(element.hasText()) result.put("text", element.ownText());
 
-                    result.put("children", getElements(element.children(), conf, errorList, log));
+                    result.put("children", getElements(element.children(), conf, errorList, log, rows, tx));
                 }
                 else {
                     if(element.hasText()) result.put("text", element.text());
                 }
+                setKernelStatusMap(tx, Map.of("rows", rows.incrementAndGet(), "result", result));
```
Comment (ncordon) [inline full/src/main/java/apoc/load/LoadHtml.java:129]: Why are we writing the result here to the kernel status map?

File: core/src/test/java/apoc/periodic/PeriodicTest.java [hunk 2]
```diff
@@ -278,6 +279,27 @@ public void testIterate() throws Exception {
                 row -> assertEquals(100L, row.get("count"))
         );
     }
+    
+    @Test
+    public void testStatusDetailsPeriodicIterate() {
+        db.executeTransactionally("UNWIND range(1,9999) AS x CREATE (:StatusIterate)");
+        KernelTestUtils.checkStatusDetails(db, 
+                "CALL apoc.periodic.iterate('match (p:StatusIterate) return p', 'SET p.lastname =p.name REMOVE p.name', {batchSize:10,parallel:true})", 
+                Collections.emptyMap(),
```
Comment (ncordon) [inline core/src/test/java/apoc/periodic/PeriodicTest.java:288]: Why is this status empty?

File: test-utils/src/main/java/apoc/kernel/KernelTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        assertEventually(() -> TestUtil.<String>singleResultFirstColumn(db,
+                "SHOW TRANSACTIONS YIELD statusDetails, currentQuery where currentQuery STARTS WITH $startQuery RETURN statusDetails", 
+                Map.of("startQuery", finalStartQuery)),
+                StringUtils::isNotEmpty, 
+                20L, TimeUnit.SECONDS);
+        
+        try {
+            future.get();
+        } catch (Exception ignored) {
+            System.out.println("ignored = " + ignored);
```
[the remark this answers is not in the document]
Comment (ncordon) [reply test-utils/src/main/java/apoc/kernel/KernelTestUtils.java:36]: What happens if the query has finished before we get to see any statistics about it?

File: core/src/test/java/apoc/periodic/PeriodicTest.java (hunk 2 above)
Comment (vga91) [reply core/src/test/java/apoc/periodic/PeriodicTest.java:288]: This `Collections.emptyMap()` is just the `params` that I pass in the `db,executeTransactionally(query, params)`

**Comment by vga91** [comment]
> I have a few suggestions:
> 
> * We are mentioning `dbms.listTransactions` here, which is deprecated, but we are using `SHOW TRANSACTIONS` (more up to date one) in the descriptions. It should be `SHOW TRANSACTIONS` everywhere.
> * Does setKernelStatus have any penalty kernel related? We are calling that sometimes on every row we process

I changed the `dbms.listTransactions` references. 
Regarding point 2, there is a slight drop in performance. Maybe we can make this `setStatusDetails` configurable. For example with a boolean, or even better with a number n to update `setStatusDetails` every `n` rows/steps?

File: test-utils/src/main/java/apoc/kernel/KernelTestUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (vga91) [reply test-utils/src/main/java/apoc/kernel/KernelTestUtils.java]: Sorry my bad, I changed it with a `throw new RuntimeException(e)`

[the remark this answers is not in the document]
Comment (ncordon) [reply test-utils/src/main/java/apoc/kernel/KernelTestUtils.java]: I still see we are ignoring all exceptions?

**Comment by ncordon** [comment]
I think it would make sense to update statistic only every so many rows. Let's say every 1000 rows. The reason for this in my opinion:
* If the transaction takes little time, you don't care about its progress.
* If the transaction is big, you normally check the progress in intervals of time so you get an idea of how much is going to take.

Can you think of an argument why we want that number of rows configurable?


If we do those changes I'm happy to approve ✅

**Comment by jexp** [comment]
@vga91 @ncordon what's the status on this one? Lots of good discussions and progress, but seems no conclusion?

**Comment by vga91** [comment]
@jexp sorry for the late.
I should have implemented Nacho's changes, but, with the CI, there seems to be some flaky tests that I missed.
I will try to fix them asap.

**Comment by vga91** [comment]
The `traceability/validate-pr` job fails because the pr is from my fork.
because the pr is from my fork.
I created the analogous pr by pushing to this repo https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/3383
@ncordon @jexp

**Pull request closed without merging**